tests/bluetooth/bsim_bt/bsim_test_audio/src/bass_client_test.c

```diff
@@ -0,0 +1,536 @@
+	case BT_GAP_LE_PHY_CODED: return "LE Coded";
+	default: return "Unknown";
+	}
+}
+
+static void bass_client_discover_cb(struct bt_conn *conn, int err,
+				    uint8_t recv_state_count)
+{
+	if (err != 0) {
+		FAIL("BASS discover failed (%d)\n", err);
+	} else {
+		printk("BASS discover done with %u recv states\n",
+		       recv_state_count);
+		g_discovery_complete = true;
+	}
+}
+
+static void bass_client_scan_cb(const struct bt_le_scan_recv_info *info,
+				uint32_t broadcast_id)
+{
+	char le_addr[BT_ADDR_LE_STR_LEN];
+
+	bt_addr_le_to_str(info->addr, le_addr, sizeof(le_addr));
+	printk("Scan Recv: [DEVICE]: %s, broadcast_id %u, "
+	       "interval (ms) %u), SID 0x%x, RSSI %i",
+	       le_addr, broadcast_id, info->interval * 5 / 4,
+	       info->sid, info->rssi);
+
+	(void)memcpy(&g_broadcaster_info, info, sizeof(g_broadcaster_info));
+	bt_addr_le_copy(&g_broadcaster_addr, info->addr);
+	g_broadcast_id = broadcast_id;
+	g_broadcaster_found = true;
+}
+
+static bool metadata_entry(struct bt_data *data, void *user_data)
+{
+	char metadata[512];
+
+	(void)bin2hex(data->data, data->data_len, metadata, sizeof(metadata));
+
+	printk("\t\tMetadata length %u, type %u, data: %s\n",
+	       data->data_len, data->type, metadata);
+
+	return true;
+}
+
+static void bass_client_recv_state_cb(struct bt_conn *conn, int err,
+				      const struct bt_bass_recv_state *state)
+{
+	char le_addr[BT_ADDR_LE_STR_LEN];
+	char bad_code[33];
+
+	if (err != 0) {
+		FAIL("BASS recv state read failed (%d)\n", err);
+		return;
+	}
+
+	bt_addr_le_to_str(&state->addr, le_addr, sizeof(le_addr));
+	(void)bin2hex(state->bad_code, BT_BASS_BROADCAST_CODE_SIZE, bad_code,
+		      sizeof(bad_code));
+	printk("BASS recv state: src_id %u, addr %s, sid %u, sync_state %u, "
+	       "encrypt_state %u%s%s\n", state->src_id, le_addr, state->adv_sid,
+	       state->pa_sync_state, state->encrypt_state,
+	       state->encrypt_state == BT_BASS_BIG_ENC_STATE_BAD_CODE ? ", bad code" : "",
+	       bad_code);
+
+	for (int i = 0; i < state->num_subgroups; i++) {
+		const struct bt_bass_subgroup *subgroup = &state->subgroups[i];
+		struct net_buf_simple buf;
+
+		printk("\t[%d]: BIS sync %u, metadata_len %u\n",
+		       i, subgroup->bis_sync, subgroup->metadata_len);
+
+		net_buf_simple_init_with_data(&buf, (void *)subgroup->metadata,
+					      subgroup->metadata_len);
+		bt_data_parse(&buf, metadata_entry, NULL);
+	}
+
+
+	if (state->pa_sync_state == BT_BASS_PA_STATE_INFO_REQ) {
+		err = bt_le_per_adv_sync_transfer(g_pa_sync, conn,
+						  BT_UUID_BASS_VAL);
+		if (err != 0) {
+			FAIL("Could not transfer periodic adv sync: %d\n", err);
+		}
+	}
+
+	g_state_synced = state->pa_sync_state == BT_BASS_PA_STATE_SYNCED;
+
+	g_src_id = state->src_id;
+	g_cb = true;
+}
+
+static void bass_client_recv_state_removed_cb(struct bt_conn *conn, int err,
+					      uint8_t src_id)
+{
+	if (err != 0) {
+		FAIL("BASS recv state removed failed (%d)\n", err);
```

Comment:
Shouldn't we exit the function after we've encountered an error?

Reply:
I guess we should. AFAIK the tests work the same regardless, but adding it for good measure :)